src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs

```diff
@@ -1192,137 +1243,180 @@ static void ProcessNode(RegexNode node, RegexNode subsequent, int maxDepth)
-                        {
-                            Debug.Assert(subsequent.Type != Group);
-                            switch (subsequent.Type)
-                            {
-                                case Concatenate:
-                                case Capture:
-                                case Atomic:
-                                case Require:
-                                case Loop when subsequent.M > 0:
-                                case Lazyloop when subsequent.M > 0:
-                                    subsequent = subsequent.Child(0);
-                                    continue;
-                            }
+            // Start by looking at the loop's sole child.
+            node = node.Child(0);
 
-                            break;
-                        }
+            // Skip past captures.
+            while (node.Type == Capture)
+            {
+                node = node.Child(0);
+            }
 
-                        // If the two nodes don't agree on case-insensitivity, don't try to optimize.
-                        // If they're both case sensitive or both case insensitive, then their tokens
-                        // will be comparable.
-                        if ((node.Options & RegexOptions.IgnoreCase) != (subsequent.Options & RegexOptions.IgnoreCase))
-                        {
-                            return false;
-                        }
+            // If the loop's body terminates with a {one/notone/set} loop, return it.
+            if (node.Type == Oneloop || node.Type == Notoneloop || node.Type == Setloop)
+            {
+                return node;
+            }
 
-                        // If the successor is an alternation, all of its children need to be evaluated, since any of them
-                        // could come after this node.  If any of them fail the optimization, then the whole node fails.
-                        if (subsequent.Type == Alternate)
-                        {
-                            int childCount = subsequent.ChildCount();
-                            for (int i = 0; i < childCount; i++)
-                            {
-                                if (!CanBeMadeAtomic(node, subsequent.Child(i), maxDepth - 1))
-                                {
-                                    return false;
-                                }
-                            }
+            // If the loop's body is a concatenate, we can skip to its last child iff that
+            // last child doesn't conflict with the first child, since this whole concatenation
+            // could be repeated, such that the first node ends up following the last.
+            if (node.Type == Concatenate)
+            {
+                int concatCount = node.ChildCount();
+                RegexNode lastConcatChild = node.Child(concatCount - 1);
+                if (CanBeMadeAtomic(lastConcatChild, node.Child(0), maxDepth - 1))
```

Comment:
In this case, we're passing in the last child of the "concat" as the node, and the first child of the "concat" in as the subsequent. Maybe subsequent is not the right term then -- referent?

Reply:
The idea is that we're testing it as if it's the subsequent node, e.g. with ab* testing what happens if it were b*a.